### ISSUE TITLE
ghprb-plugin error + DataBoundConfigurator for Unclassified GlobalConfigurationCategoryConfigurator questions

### ISSUE THREAD

**Issue by jcrben** [jenkinsci/configuration-as-code-plugin#418]
Hey - I've been asking questions around debugging in the Gitter channel and but have now hit my "open a bug" point. Pardon my ignorance - I'm pretty much clueless with Java.

Reproduction repo is based upon https://gitlab.com/jcrben-repro/jenkins-jcasc-docker - if you want to a full repro, I can do it (but it'd take time since my real repo is private) - I think there's enough to go by here.

Using ghprbtrigger results in this message:
![image](https://user-images.githubusercontent.com/5614134/43698824-cea949d8-9900-11e8-8856-26a559826cdb.png)

yaml file looks something like this:
```
jenkins:
  authorizationStrategy:
    loggedInUsersCanDoAnything:
      allowAnonymousRead: false
  securityRealm:
    local:
      allowsSignup: false
      users:
        - id: "admin"
          password: "4321"
unclassified:
  ghprbtrigger:
    manageWebhooks: false
    adminList: 'blahblahblah'
    useGitHubHooks: true
```

## Investigatory musings
The ghprb-plugin has a bunch of global configuration that I'd like to access at https://github.com/jenkinsci/ghprb-plugin/blob/4e86ed47a96a01eeaa51a479ff604252109635f6/src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java#L175-L202 and it already has a DataBoundConstructor on it as well as a detailed global.groovy to pass that check. 
<img width="1287" alt="screen shot 2018-08-05 at 10 38 39 pm" src="https://user-images.githubusercontent.com/5614134/43698750-729e10b0-9900-11e8-8aab-6229413a0ec6.png">

Also it has a [config.groovy](https://github.com/jenkinsci/ghprb-plugin/blob/4e86ed47a96a01eeaa51a479ff604252109635f6/src/main/resources/org/jenkinsci/plugins/ghprb/GhprbTrigger/config.groovy) - not sure how important that is.

In the describe for the GlobalConfigurationCategoryConfigurator, it uses [this stream](https://github.com/jenkinsci/configuration-as-code-plugin/blob/2aff7e32de393848cb10908a4b2d7cc679595bd1/plugin/src/main/java/org/jenkinsci/plugins/casc/impl/configurators/GlobalConfigurationCategoryConfigurator.java#L69-L77) and maps over everything into a DescriptorConfigurator and I wonder if that might be related to the issue. The attribute for ghprbtrigger is (`ghprbtrigger`, `class hudson.maven.GithubTrigger$DescriptorImpl`) and it sort of seems like what I want is `class hudson.maven.GithubTrigger$GithubTrigger` (the DataBoundConstructor constructor function).

My investigation: I've been stepping thru the code with Intellij and the [BaseConfigurator configure block](https://github.com/jenkinsci/configuration-as-code-plugin/blob/2aff7e32de393848cb10908a4b2d7cc679595bd1/plugin/src/main/java/org/jenkinsci/plugins/casc/BaseConfigurator.java#L243-L306) is the area I'm focused on. After `describe()` (which is where the above happens), there's eventually this [`lookup()`](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/java/org/jenkinsci/plugins/casc/Configurator.java#L94). I would like to match with the DataBoundConstructor in that lookup, but since it's checking this `class hudson.maven.GithubTrigger$DescriptorImpl` method instead of the constructor, that isn't happening. It's also returning a DescriptorConfigurator before DataBoundConfigurator so I swapped the order of those two in my fork.

I suspect the `@Extension` annotation sitting above `class hudson.maven.GithubTrigger$DescriptorImpl` and not the constructor might also have something to do with what's happening here also...

**Comment by ndeloof** [comment]
`GithubTrigger$DescriptorImpl` (like many Descriptors) is implemented with manually written json data-binding code https://github.com/jenkinsci/ghprb-plugin/blob/4e86ed47a96a01eeaa51a479ff604252109635f6/src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java#L899

Configuration-as-Code do rely on declarative data-binding so it can introspect the data model, i.e we need `set` methods for each configurable attribute (preferably `@DataBoundSetter`s). I expect in near future to see much plugins to adopt such setters and replace hand-written code with one-liner `req.bindJson(this, formData)`, but in the meantime such components are not supported - until you write some custom code for it, but then simpler to just fix the plugin.

**Comment by ndeloof** [comment]
to clarify, CasC let you configure persistent data of the Descriptor (i.e global.jelly) but not for GithubTrigger instance, which are job-level components. Check job-dsl plugin if you want to configure such element.

**Comment by jcrben** [comment]
Thanks for your advice.

To clarify, I'm interested in the global configuration - the stuff listed at https://github.com/jenkinsci/ghprb-plugin/blob/b0c6281cdf1600abc098b3fe320472d57e12acee/src/main/resources/org/jenkinsci/plugins/ghprb/GhprbTrigger/global.groovy or in the global Configure System UI:
![image](https://user-images.githubusercontent.com/5614134/43942718-edf28cde-9c2d-11e8-939f-700b16821b22.png)

It looks like the single config being picked up is this:
![image](https://user-images.githubusercontent.com/5614134/43942763-17a2c2c4-9c2e-11e8-97e9-5609c623abcd.png)

And that has a setter, so I tried adding a setManageWebhooks setter - something like this:
```
public void setManageWebhooks(boolean setManageWebhooks) {
            this.manageWebhooks = setManageWebhooks;
}
```

But then `mvn install` yelled at me about `manageWebhooks` so I tried declaring it at the class level  and saw `JSONException: JSONObject["manageWebhooks"] is not a Boolean.` - tried a couple other things too. I'll step into it in a debugger at some point and see if I can fumble my way further...

Basically would just like to expose these global things; noticed https://github.com/jenkinsci/github-plugin/blob/497f43251e3bfc130f125f75e45f73e06894361d/src/main/java/org/jenkinsci/plugins/github/config/GitHubPluginConfig.java which has that `req.bindJson` thing https://github.com/jenkinsci/github-plugin/blob/497f43251e3bfc130f125f75e45f73e06894361d/src/main/java/org/jenkinsci/plugins/github/config/GitHubPluginConfig.java#L164

**Comment by ndeloof** [comment]
JCasC will indeed require ghprdb-plugin to change https://github.com/jenkinsci/ghprb-plugin/blob/master/src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java#L899 so it relies on `bindJson` and expose (@DataBound)Setters to introspection and programmatic configuration.

**Comment by jcrben** [comment]
Cool, thanks again - I did enough to get something to show up in the docs  https://github.com/jcrben/ghprb/commit/1dc061b24acca291245c0ff563d55821a4323252 and opened the above issue in the plugin's repo - maybe I'll try to chip away at it

**Comment by wosc** [comment]
Hurray, someone started implementing this in https://github.com/jenkinsci/ghprb-plugin/pull/731
